impala/tests/test_hs2_fault_injection.py

```diff
@@ -0,0 +1,364 @@
+        self.transport = FaultInjectingHttpClient(url)
+        self.configuration = {'idle_session_timeout': '30'}
+
+    def teardown(self):
+        self.transport.disable_fault()
+
+    def connect(self):
+        self.transport.open()
+        protocol = TBinaryProtocol(self.transport)
+        service = None
+        if six.PY2:
+            # ThriftClient == ImpalaHiveServer2Service.Client
+            service = ThriftClient(protocol)
+        elif six.PY3:
+            # ThriftClient == TClient
+            # service = ThriftClient(protocol)
+            service = ThriftClient(ImpalaHiveServer2Service, protocol)
+        service = HS2Service(service, retries=3)
+        return hs2.HiveServer2Connection(service, default_db=None)
+
+    def __expect_msg_retry(self, impala_rpc_name):
+        """Returns expected log message for rpcs which can be retried"""
+        return ("Caught HttpError HTTP code 502: Injected Fault  in {0} (tries_left=3)".
+                format(impala_rpc_name))
+
+    def __expect_msg_retry_with_extra(self, impala_rpc_name):
+        """Returns expected log message for rpcs which can be retried and where the http
+        message has a message body"""
+        return ("Caught HttpError HTTP code 503: Injected Fault EXTRA in {0} (tries_left=3)".
+                format(impala_rpc_name))
+
+    def __expect_msg_retry_with_retry_after(self, impala_rpc_name):
+        """Returns expected log message for rpcs which can be retried and where the http
+        message has a body and a Retry-After header that can be correctly decoded"""
+        return ("Caught HttpError HTTP code 503: Injected Fault EXTRA in {0} (tries_left=3), retry after 1 secs".
+                format(impala_rpc_name))
+
+    def __expect_msg_retry_with_retry_after_sleep(self):
+        """Returns expected log message for the sleep which uses a value
+        from the Retry-After header"""
+        return ("sleeping after seeing Retry-After value of 1")
+
+    def __expect_msg_retry_after_default_sleep(self):
+        """Returns expected log message for the default sleep time of 1 second"""
+        return ("sleeping for 1 second before retrying")
+
+    def __expect_msg_retry_with_retry_after_no_extra(self, impala_rpc_name):
+        """Returns expected log message for rpcs which can be retried and the http
+        message has a Retry-After header that can be correctly decoded"""
+        return ("Caught HttpError HTTP code 503: Injected Fault  in {0} (tries_left=3), retry after 1 secs".
+                format(impala_rpc_name))
+
+    def __expect_msg_no_retry(self, impala_rpc_name):
+        """Returns expected log message for rpcs which can not be retried"""
+        return ("Caught HttpError HTTP code 502: Injected Fault  in {0} which is not retryable".
+                format(impala_rpc_name))
+
+    def test_connect(self, caplog):
+        """Tests fault injection in connect().
```

Comment:
I could be wrong here, but I think the fault is injected in cursor() call. OpenSession rpc is called from cursor().

Reply:
You're right. Done.